### ISSUE TITLE
Weird type assertion rules in TypeScript

### ISSUE THREAD

**Issue by SergeyTeplyakov** [Microsoft/TypeScript#20849]
**TypeScript Version:**  recent master

Consider the following case:

```ts
interface I {a: string; b: string;}

// No errors
const x = <I>{a: "1", b: 'sdf', z: 42};

// Error: Object literal may specify only known properties
const x2: I = {a: "1", b: 'sdf', z: 42};

// Error: property 'b' is missing
const x3 = <I>{a: '1', z: 42};

// OK
const x4 = <I>{a: '1'};
```

So I can't get the assigning rules in TypeScript:

1. Why the first case with typecast is fine, but the second case with explicit type argument is not? I known that the excessive properties check was added recently, but I can't see why it is not enabled for cast operator as well?
2. The last two cases are even weirder: why the `<I>{a: '1'}` is ok, but `<I>{a: '1', unknownProp: 2}` is not?

**Comment by DanielRosenwasser** [comment]
The type assertion indicates "I know what I am doing, so please take some measures to coerce the type and don't perform these checks", but we still have some guard-rails in place. Currently, the guard-rail is a type relationship (called "comparability") that more or less checks that for two object types, one must be a superset of the other's properties. The idea is we don't want users to accidentally cast to unrelated types.

Unfortunately, this means the type system won't allow type assertions on the object type lattice that aren't reachable in a consistent upward or downward direction. Check out the following to see what I mean by this:

```
                       { a: string }
                        /          \
{ a: string, b: string }            { a: string, z: number }
                        \          /
            { a: string, b: string, z: number }
```

____

> why the `<I>{a: '1'}` is ok, but `<I>{a: '1', unknownProp: 2}` is not?

For `x4`, you are moving down the object lattice consistently, but for `x3`, the rule is that the type of `I` (i.e. `{ a: string, b: string }`) must be comparable to `{ a: string, z: number }`, or vice-versa. For `x3`, `I` has no `z` property, and `{ a: '1', z: 42 }` has no `b` property. If `I` was instead defined as `{ a: string, b: number, z: number }`, then it'd work.

**Comment by smera** [comment]
I think I understand the need to put some guardrails in place, and that comparability is a required condition to make sure the cast is not completely off (so if A and B are not comparable, then casting A to B never makes sense). But wondering why assignability can't be used instead as the guardrail, which is a stricter condition (so casting a value of type A to B would be allowed if A is assignable to B), and therefore could catch a larger set of issues,

**Comment by DanielRosenwasser** [comment]
[Comparability](https://github.com/Microsoft/TypeScript/blob/2fb84d5fc5227ed33f578a358ce5e63833445370/doc/spec/Types.md#comparability--comparability-) is mainly more lax on unions in a few ways that make it easier to work with in places where two types could plausibly be the same at runtime.

The original motivating example was comparing `number | "hello"` to `string`. We needed to be able to check using `===` whether something of type `number | "hello"` is equal to a value of type `string`. But `string` is not assignable to `number | "hello"` (`string` isn't assignable to either), and the opposite is not true either (since `number` is not assignable to `string).

Instead of checking whether *every* constituent of a union is related to the other type, comparability checks whether any given constituent of a union is related to the other. This actually went a long way, but some of the same problems you're describing existed in the assignability model and still carried over.

Obviously it's a balancing act between ease of use and catching mistakes. If we had a different relationship that was easy to reason about, I think we'd consider it. We recently did, but it was sometimes *too* lax (allowing you to compare an array to a string because they both have a `length` property of type `number`).
